### PR TITLE
Store UGID in group cache to handle group renames at provider's end 

### LINKED ISSUE #620: Issue: EntraID groups that are renamed result in auth failures because GID conflicts
### Is there an existing issue for this?

- [X] I have searched the existing issues and found none that matched mine

### Describe the issue

Authd appears to cache the name of each group a user is a member of, as supplied by the OIDC claim, with a randomly generated GID.  However, group names can be changed at any time within EntraID. I'm not sure what is actually included in the claim, but if I explicitly rename a group, after having previously logged in to a server, the new group name immediately starts causing a conflict. This only impacts password authentication -- I can still SSH using a key -- but there is no way a user can resolve this without admin intervention.

User experience looks like this:

```
$ ssh first.last@domain.com@192.168.1.1
Insert 'r' to cancel the request and go back
(first.last@domain.com@192.168.1.1) Enter your local password:
authentication status failure: can't check authentication: failed to update user "first.last@domain.com": GID for group "NEW_GROUP" already in use by a different group
Insert 'r' to cancel the request and go back
(first.last@domain.com@192.168.1.1) Enter your local password:
```

The log files show the following entries:
```
2024-11-06T23:22:37.528555+00:00 analysis authd[14833]: 2024/11/06 23:22:37 ERROR GID 1499969210 for group "NEW_GROUP" already in use by group "OLD_GROUP"
2024-11-06T23:22:37.533334+00:00 analysis authd[14833]: 2024/11/06 23:22:37 WARN can't check authentication: failed to update user "first.last@domain.com": GID for group "NEW_GROUP" already in use by a different group
```

This work-around seems to work:
* Remove the user from the EntraID group
* Log into the authd-enabled server, then log out
* Add the user back to the EntraID group
* Log back into the authd-enabled server

### Steps to reproduce

* Add a user to an EntraID group
* Log into an authd-enabled server, then log out
* Renamed the EntraID group
* Attempt to log back into authd-enabled server, observe GID conflict error

#### authd version
```
authd   0.3.6
```

#### authd-msentraid broker version
```
name:      authd-msentraid
summary:   MSEntra ID broker for authd
publisher: Canonical**
store-url: https://snapcraft.io/authd-msentraid
license:   GPL-3.0
description: |
  This is the MS Entra ID broker snap for authd  to provide MS Entra ID OIDC
  based authentication on Ubuntu with authd.
services:
  authd-msentraid: simple, enabled, active
snap-id:      vS3oJLMss6lgWwoFcPqYDUA2HB20I1Dc
tracking:     0.x/edge
refresh-date: today at 01:15 UTC
channels:
  0.x/stable:    0.1+4fe9826.0f76acc 2024-10-02 (51) 18MB -
  0.x/candidate: ^
  0.x/beta:      ^
  0.x/edge:      0.1+fef0c13.918f3f8 2024-11-05 (61) 18MB -
installed:       0.1+fef0c13.918f3f8            (61) 18MB -
```

#### gnome-shell version
```
gnome-shell:
  Installed: 46.3.1-1ubuntu1~24.04.1authd2
  Candidate: 46.3.1-1ubuntu1~24.04.1authd2
  Version table:
 *** 46.3.1-1ubuntu1~24.04.1authd2 500
        500 https://ppa.launchpadcontent.net/ubuntu-enterprise-desktop/authd/ubuntu noble/main amd64 Packages
        100 /var/lib/dpkg/status
     46.0-0ubuntu6~24.04.5 500
        500 http://us.archive.ubuntu.com/ubuntu noble-updates/main amd64 Packages
     46.0-0ubuntu6~24.04.3 500
        500 http://security.ubuntu.com/ubuntu noble-security/main amd64 Packages
     46.0-0ubuntu5 500
        500 http://us.archive.ubuntu.com/ubuntu noble/main amd64 Packages
```

#### Distribution
```
Distributor ID: Ubuntu
Description:    Ubuntu 24.04.1 LTS
Release:        24.04
Codename:       noble
```

#### Logs
```
[180724.694978] analysis authd[14833]: 2024/11/06 23:22:37 ERROR GID 1499969210 for group "NEW_GROUP" already in use by group "OLD_GROUP"
[180724.694978] analysis authd[14833]: 2024/11/06 23:22:37 WARN can't check authentication: failed to update user "first.last@domain.com": GID for group "NEW_GROUP" already in use by a different group
```

#### /etc/authd/brokers.d/msentraid.conf
```
# This section is used by authd to identify and communicate with the broker.
# It should not be edited.
[authd]
name = Microsoft Entra ID
brand_icon = /snap/authd-msentraid/current/broker_icon.png
dbus_name = com.ubuntu.authd.MSEntraID
dbus_object = /com/ubuntu/authd/MSEntraID
```

#### authd-msentraid configuration
```
[oidc]
issuer = https://login.microsoftonline.com/<UUID redacted>/v2.0
client_id = <UUID redacted>

[users]
# The directory where the home directory will be created for new users.
# Existing users will keep their current directory.
# The user home directory will be created in the format of {home_base_dir}/{username}
home_base_dir = /home

# The username suffixes that are allowed to login via ssh without existing previously in the system.
# The suffixes must be separated by commas.
# ssh_allowed_suffixes = @example.com,@anotherexample.com
ssh_allowed_suffixes = @<redacted>
```

### Double check your logs

- [X] I have redacted any sensitive information from the logs

### PR DESCRIPTION
**Pull request by shiv-tyagi**
Closes #620
UDENG-5160

As suggested by @adombeck, to identify if a group is renamed at provider's end, we should store UGID along with the GID and group name in the cache and find if there is already a group record with same UGID in cache. In case we find one, we delete the record being pointed by old name in relevant  buckets and populate with record containing new name.

### CODE REVIEW

**Comment by shiv-tyagi** [comment]
@adombeck Here is my initial work. Could you please check if I am going in right direction?

**Comment by adombeck** [comment]
> I saw the test failures and found that since we have added the UGID field, we need to update the test data files almost everywhere. There are a lot of them.

Ah, I assume it's the golden files which need to be updated. You can update them all at once by running `TESTS_UPDATE_GOLDEN=1 go test ./...`.

**Comment by shiv-tyagi** [comment]
> `TESTS_UPDATE_GOLDEN=1 go test ./...`.

I ran this and a lot of golden files did get updated, but the *.db files in the test data like [this one](internal/services/pam/testdata/TestIDGeneration/golden/generate_id/cache.db) still couldn't get updated. Do I need to update them manually or am I missing something?

**Comment by adombeck** [comment]
> I ran this and a lot of golden files did get updated, but the *.db files in the test data like [this one](https://github.com/ubuntu/authd/pull/internal/services/pam/testdata/TestIDGeneration/golden/generate_id/cache.db) still couldn't get updated. Do I need to update them manually or am I missing something?

Those should also be updated automatically. I ran `TESTS_UPDATE_GOLDEN=1 go test ./internal/services/pam/...` locally on your branch and it updated a lot of files, including the `cache.db` files.

**Comment by shiv-tyagi** [comment]
I get this error when I do run the command

> Those should also be updated automatically. I ran `TESTS_UPDATE_GOLDEN=1 go test ./internal/services/pam/...` locally on your branch and it updated a lot of files, including the `cache.db` files.

```# github.com/ubuntu/authd/internal/users/localgroups
internal/users/localgroups/localgroups.go:21:16: undefined: getPasswdUsernames
FAIL    github.com/ubuntu/authd/internal/services/pam [build failed]
FAIL```

I get this error when I run the command.

**Comment by shiv-tyagi** [comment]
> Those should also be updated automatically. I ran `TESTS_UPDATE_GOLDEN=1 go test ./internal/services/pam/...` locally on your branch and it updated a lot of files, including the `cache.db` files

Sorry for bugging you so much. I was finally able to run tests and get the cache.db files updated as well. But I still couldn't get [these *.db.yaml files](https://github.com/ubuntu/authd/blob/315214ad65efe0cf424a227bbb0248a569c3d831/internal/users/cache/testdata/multiple_users_and_groups.db.yaml) updated. Did yours got updated when you ran it locally?

**Comment by 3v1n0** [comment]
Just use `TESTS_UPDATE_GOLDEN=1 go test -C ./internal ./...` it should work

**Comment by adombeck** [comment]
> internal/users/localgroups/localgroups.go:21:16: undefined: getPasswdUsernames
> FAIL    github.com/ubuntu/authd/internal/services/pam [build failed]
> FAIL
> 
> I get this error when I run the command. Anything I am missing?

Do you have `CGO_ENABLED` set to `0`? If so, unset it please:

```
CGO_ENABLED= TESTS_UPDATE_GOLDEN=1 go test ./internal/services/pam/...
```

**Comment by shiv-tyagi** [comment]
Thanks @3v1n0 and @adombeck, I was able to run the tests and get the golden files updated. There's this one QA and Sanity check failing. I checked the logs but couldn't find any error message. Any idea how to fix that?

**Comment by adombeck** [comment]
> Thanks @3v1n0 and @adombeck, I was able to run the tests and get the golden files updated. There's this one QA and Sanity check failing. I checked the logs but couldn't find any error message. Any idea how to fix that?

That job unfortunately doesn't print its results on the job page, but on the summary page: https://github.com/ubuntu/authd/actions/runs/12152788489/attempts/1?pr=647

image

**Comment by adombeck** [comment]
To catch linter errors earlier, you could install a git pre-push hook which runs [golangci-lint](https://github.com/golangci/golangci-lint) locally:

```bash
$ cat .git/hooks/pre-push
#!/bin/sh

set -eu
set -x

golangci-lint run
```

**Comment by shiv-tyagi** [comment]
Thanks again @adombeck for patiently answering my questions. I am opening the PR for review. 

Last few questions (I promise), 
- Is there a chat server (discord or slack etc.) where the developers hang out and I can quickly ask my doubts there?
- Are there any regular online meetings which I can join to learn and get involved more into the project?

**Comment by adombeck** [comment]
> Thanks again @adombeck for patiently answering my questions. I am opening the PR for review. 

No problem :)

> Last few questions - I promise :), 
> - Is there a chat server (discord or slack etc.) where the developers hang out and I can quickly ask my doubts?
> - Are there any regular online meetings which I can join to learn and get involved more in the project?

Currently, we don't have a public chat server or public meetings. I'll mention to the team that you're interested in that, but I can't promise anything 🫤

File: internal/users/manager.go [hunk 1]
```diff
@@ -132,7 +134,7 @@ func (m *Manager) UpdateUser(u UserInfo) (err error) {
 			localGroups = append(localGroups, g.Name)
```
[the remark this answers is not in the document]
Comment (shiv-tyagi) [reply internal/users/manager.go:134]: Done!

File: internal/users/cache/update.go [hunk 2]
```diff
@@ -92,14 +92,26 @@ func updateGroups(buckets map[string]bucketWithName, groupContents []GroupDB) er
...
 
 		// If a group with the same GID exists, we need to ensure that it's the same group or fail the update otherwise.
-		if existingGroup.Name != "" && existingGroup.Name != groupContent.Name {
-			log.Errorf(context.TODO(), "GID %d for group %q already in use by group %q", groupContent.GID, groupContent.Name, existingGroup.Name)
+		if existingGroup.UGID != "" && existingGroup.UGID != groupContent.UGID {
+			log.Errorf(context.TODO(), "GID %d for group with UGID %q already in use by a group with UGID %q", groupContent.GID, groupContent.UGID, existingGroup.UGID)
 			return fmt.Errorf("GID for group %q already in use by a different group", groupContent.Name)
 		}
 
+		// Same UID and UGID but a different Name can happen due to group renaming at provider's end.
```
[the remark this answers is not in the document]
Comment (shiv-tyagi) [reply internal/users/cache/update.go]: Fixed.

File: internal/users/cache/update.go [hunk 3]
```diff
@@ -92,14 +92,26 @@ func updateGroups(buckets map[string]bucketWithName, groupContents []GroupDB) er
 		}
 
 		// If a group with the same GID exists, we need to ensure that it's the same group or fail the update otherwise.
-		if existingGroup.Name != "" && existingGroup.Name != groupContent.Name {
-			log.Errorf(context.TODO(), "GID %d for group %q already in use by group %q", groupContent.GID, groupContent.Name, existingGroup.Name)
+		if existingGroup.UGID != "" && existingGroup.UGID != groupContent.UGID {
```
[the remark this answers is not in the document]
Comment (shiv-tyagi) [reply internal/users/cache/update.go]: As soon as I removed this check, we start getting the following errors on running tests:-
```
2024/12/15 17:14:13 ERROR GID 11111 for group with UGID "12345678" already in use by a group with UGID ""
2024/12/15 17:14:13 ERROR GID 22222 for group with UGID "56781234" already in use by a group with UGID ""
```

Looks like the place from where the test code is picking the group metadata does not have UGID set.
Can you please point me to the files where I need to update the UGID?

**Comment by shiv-tyagi** [comment]
@adombeck Sorry it is taking longer because I am getting less time after work. I only get weekends to contribute.
I look forward to get this PR merged soon. I have asked some questions in the comments above.

File: internal/users/cache/update.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (adombeck) [reply internal/users/cache/update.go]: I pushed a fix

File: internal/users/cache/update.go [hunk 4]
```diff
@@ -92,14 +92,26 @@ func updateGroups(buckets map[string]bucketWithName, groupContents []GroupDB) er
...
+			log.Errorf(context.TODO(), "GID %d for group with UGID %q already in use by a group with UGID %q", groupContent.GID, groupContent.UGID, existingGroup.UGID)
 			return fmt.Errorf("GID for group %q already in use by a different group", groupContent.Name)
 		}
 
+		// Same UID and UGID but a different Name can happen due to group renaming at provider's end.
+		if existingGroup.Name != "" && existingGroup.Name != groupContent.Name {
```
[the remark this answers is not in the document]
Comment (adombeck) [reply internal/users/cache/update.go]: I assume the intend was to only do this check if an existing group with the same UID was found. I pushed a commit which does that and still handles the corner case that the existing group has an empty name correctly.

**Review by adombeck — APPROVED** [review]
This looks good to me now, thanks a lot @shiv-tyagi!

**Comment by adombeck** [comment]
@denisonbarbosa @3v1n0 does one of you want to take a look before I merge?

**Comment by shiv-tyagi** [comment]
Thanks @adombeck.

**Comment by adombeck** [comment]
> @denisonbarbosa @3v1n0 does one of you want to take a look before I merge?

@denisonbarbosa @3v1n0 ping

File: internal/users/cache/update.go [hunk 5]
```diff
@@ -93,15 +93,16 @@ func updateGroups(buckets map[string]bucketWithName, groupContents []GroupDB) er
 		if err != nil && !errors.Is(err, NoDataFoundError{}) {
 			return err
 		}
+		groupExists := !errors.Is(err, NoDataFoundError{})
```
Comment (3v1n0) [inline internal/users/cache/update.go:96]: A nit, but I think it becomes clearer with:

```go
		groupExists := !errors.Is(err, NoDataFoundError{})
		if err != nil && groupExists {
			return err
		}
```

**Review by 3v1n0 — APPROVED** [review]
LGTM, just a [nit](https://github.com/ubuntu/authd/pull/647#discussion_r1892935317) to make things a bit more readable.

**Pull request merged**